src/elm/Types/DataTrigger.elm

```diff
@@ -255,14 +280,29 @@ decoder =
         |> resolve
 
 
-toDecoder : String -> Int -> DataTriggerEvent -> String -> Operator -> Maybe ( String, JsonType ) -> Decoder DataTrigger
-toDecoder iName iMajor on path operator maybeKnownValue =
-    case maybeKnownValue of
-        Just ( value, jType ) ->
-            Decode.succeed (DataTrigger iName iMajor on path operator value jType)
+toDecoder : String -> String -> String -> Int -> DataTriggerEvent -> String -> Operator -> Maybe ( String, JsonType ) -> Decoder DataTrigger
+toDecoder groupName deviceId iName iMajor on path operator maybeKnownValue =
+    let
```

Comment:
didn't test it, but this is the way to go
```
let
    resultValue =
        case maybeKnownValue of
            Just ( value, jType ) ->
                Ok (value, jType)

            Nothing ->
                Err "some error....."

    resultTarget =
        case ( groupName, deviceId ) of
            ( "*", "*" ) ->
                Ok AllDevices

            ( _, "*" ) ->
                Ok (DeviceGroup groupName)

            ( "*", _ ) ->
                Ok (SpecificDevice deviceId)

            ( _, _ ) ->
                Err "Cannot have both device ID and group name"
in
case ( resultValue, resultTarget ) of
    ( Ok (knownValue, knownValueType), Ok target ) ->
        Decode.succeed (DataTrigger iName iMajor on path operator knownValue knownValueType target)

    ( Err msg, _ ) ->
        Decode.fail msg

    ( _, Err msg ) ->
        Decode.fail msg
```

Reply:
Nice, that's more like it. I admit I struggled a bit with Elm not knowing how to properly exploit conditionals/pattern matching.
I directly copied your snippet and it works fine, thanks!